README.md
# github-action-psscriptanalyzer

[![GitHub Actions Status][github-actions-badge]][github-actions-build]

[GitHub Action](https://github.com/features/actions) to run [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) static code analysis checks on PowerShell for [Pull Requests](https://help.github.com/articles/about-pull-requests/).

## Success Criteria

By default, this action will succeed if **zero** PSScriptAnalyzer **errors** and **warnings** are found.
Failing on errors, warnings, or informational issues can be configured. See [Usage](#Usage) below.
The sending of comments back to the PR if the action fails can be disabled if desired.

## Usage

### Basic

Basic configuration that will run PSSA and fail on errors or warnings, and send a comment back to the PR with a summary.
Note, that `repoToken` is required for sending comments back.

```yaml
name: CI
on: [pull_request]
jobs:
  lint:
    name: Run PSSA
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: lint
      uses: devblackops/github-action-psscriptanalyzer@master
      with:
        repoToken: ${{ secrets.GITHUB_TOKEN }}
```

### Advanced

Advanced configuration that will run PSSA only in the `MyModule` directory, with custom PSSA settings, and fail on errors, warnings, or informational issues.
A comment back to the PR with the PSSA summary will also be sent if any issues were detected.

```yaml
name: CI
on: [pull_request]
jobs:
  lint:
    name: Run PSSA
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: lint
      uses: devblackops/github-action-psscriptanalyzer@master
      with:
        rootPath: MyModule
        settingsPath: pssa_settings.psd1
        sendComment: true
        repoToken: ${{ secrets.GITHUB_TOKEN }}
        failOnErrors: true
        failOnWarnings: true
        failOnInfos: true
```

### Docker

Use the Docker Hub version of the Action instead of building the container during the check.

```yaml
name: CI
on: [pull_request]
jobs:
  lint:
    name: Run PSSA
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: lint
      uses: docker://devblackops/github-action-psscriptanalyzer:2.3.0
      with:
        repoToken: ${{ secrets.GITHUB_TOKEN }}
```

## Inputs

| Name | Default | Description |
|------|---------|-------------|
| rootPath       | \<none>  | The root directory to run PSScriptAnalyzer on. By default, this is the root of the repository.
| settingsPath   | \<none>  | The path to a PSScriptAnalyser settings file to control rules to execute.
| sendComment    | true  | Enable/disable sending comments with PSScriptAnalyzer results back to PR.
| repoToken      | \<none>  | GitHub token the action will use to send comments back to PR with. Use `${{ secrets.GITHUB_TOKEN }}`.
| failOnErrors   | true  | Enable/disable failing the action on PSScriptAnalyzer error items.
| failOnWarnings | true  | Enable/disable failing the action on PSScriptAnalyzer warning items.
| failOnInfos    | false | Enable/disable failing the action on PSScriptAnalyzer informational items.

## Example

![](media/example.png)

[github-actions-badge]: https://github.com/devblackops/github-action-psscriptanalyzer/workflows/CI/badge.svg
[github-actions-build]: https://github.com/devblackops/github-action-psscriptanalyzer/actions


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.4.0] - 2021-10-22

### Changed

- [**#PR18**](https://github.com/devblackops/github-action-psscriptanalyzer/pull/18) - Updated pwsh, Ubuntu, and PSSA versions (via [@alagoutte](https://github.com/alagoutte))
  - pwsh `7.0.3` -> `7.1.5`
  - Ubuntu `18.04` -> `20.04`
  - PSSA `1.19.1` -> `1.20.0`

## [2.3.1] - 2021-02-27

### Fixed

- [**#PR15**](https://github.com/devblackops/github-action-psscriptanalyzer/pull/15) - Typo in code comment (via [@revolter](https://github.com/revolter))

## [2.3.0] - 2020-08-31

### Fixed

- [**#PR13**](https://github.com/devblackops/github-action-psscriptanalyzer/pull/13) - Remove `outputs` section from `action.yml` to resolve error when `outputs` is empty (via [@SleepySysadmin](https://github.com/SleepySysadmin))

### Changed

- Updated docker image to `powershell:7.0.3-ubuntu-18.04`.

- Updated `PSScriptAnalyzer` to `1.19.1`.

## [2.2.0] - 2020-06-30

### Fixed

- [**#PR8**](https://github.com/devblackops/github-action-psscriptanalyzer/pull/8) - Fix SendComment (Only if RepoToken is available) (via [@alagoutte](https://github.com/alagoutte))

### Changed

- [**#PR10**](https://github.com/devblackops/github-action-psscriptanalyzer/pull/10) - Update checkout action to v2 (`actions/checkout@v2`) (via [@alagoutte](https://github.com/alagoutte))

- [**#PR7**](https://github.com/devblackops/github-action-psscriptanalyzer/pull/7) - Updated PSScriptAnalyzer to `1.19.0` (via [@alagoutte](https://github.com/alagoutte))

- [**#PR6**](https://github.com/devblackops/github-action-psscriptanalyzer/pull/6) - Updated PowerShell Docker iamge to `7.0.1` (via [@alagoutte](https://github.com/alagoutte))

## [2.1.1] - 2020-05-15

### Fixed

- Fixed reference to warning environment variable (`INPUT_FAILONWARNINGS`) (via [@CrazyCodeUK](https://github.com/CrazyCodeUK))

## [2.1.0] - 2020-05-15

### Changed

- Bumped PSScriptAnalyzer to `1.18.3` (via [@alagoutte](https://github.com/alagoutte))

## [2.0.0] - 2019-08-28

### Changed

- Refactored to new GitHub Action syntax
- Renamed options to:
  - rootPath
  - settingsPath
  - repoToken
  - sendComment
  - failOnErrors
  - failOnWarnings
  - failOnInfos
- Allow failing the GitHub Action on PSSA error, warning, or informational issues

## [1.2.1] - 2019-01-15

### Fixed

- Action will now only attempt to post a PR comment if a valid comments URL is found in the action payload

## [1.2.0] - 2019-01-14

### Changed

- Reorganize project so it is compatible with GitHub Marketplace

## [1.1.0] - 2018-01-06

### Changed

- Call PowerShell entrypoint script directly instead of calling bash first.
- Allow disabling sending PSScriptAnalyzer results back to PR by setting the environment variable `PSSCRIPTANALYZER_SEND_COMMENT` to `false` or `0`.

## [1.0.1] - 2018-12-28

### Fixed

- Fix logic and only post a GitHub comment to the PR if issues were found.

## [1.0.0] - 2018-12-28

### Added

- Initial release using PSScriptAnalyzer version `1.17.1`


build.ps1
[cmdletbinding(DefaultParameterSetName = 'task')]
param(
    [parameter(ParameterSetName = 'task', Position = 0)]
    [string[]]$Task = 'default',

    [parameter(ParameterSetName = 'help')]
    [switch]$Help,

    [switch]$Bootstrap
)

$ErrorActionPreference = 'Stop'

# Bootstrap dependencies
if ($Bootstrap.IsPresent) {
    Get-PackageProvider -Name Nuget -ForceBootstrap | Out-Null
    Set-PSRepository -Name PSGallery -InstallationPolicy Trusted
    if (-not (Get-Module -Name PSDepend -ListAvailable)) {
        Install-Module -Name PSDepend -Repository PSGallery -Scope CurrentUser
    }
    Import-Module -Name PSDepend -Verbose:$false
    Invoke-PSDepend -Path './requirements.psd1' -Install -Import -Force -WarningAction SilentlyContinue
}

# Execute psake task(s)
$psakeFile = './psakeFile.ps1'
Set-BuildEnvironment -Force -WarningAction SilentlyContinue
if ($Help.IsPresent) {
    Get-PSakeScriptTasks -buildFile $psakeFile  |
        Format-Table -Property Name, Description, Alias, DependsOn
} else {
    Invoke-psake -buildFile $psakeFile -taskList $Task -nologo -Verbose:($VerbosePreference -eq 'Continue')
    exit ([int](-not $psake.build_success))
}


analyze/run.ps1
#requires -modules PSScriptAnalyzer

[cmdletbinding()]
param()

$ErrorActionPreference = 'Stop'

$nl = [Environment]::NewLine

$analyzeParams = @{
    Recurse = $true
}

# By default, run PSScriptAnalyzer on the whole repository
# but allow overriding this with PSSCRIPTANALYZER_ROOT environment variable
if ($env:PSSCRIPTANALYZER_ROOT) {
    $analyzeParams.Path = $env:PSSCRIPTANALYZER_ROOT
} else {
    $analyzeParams.Path = $env:GITHUB_WORKSPACE
}

# Path to custom script analzyer settings
if ($env:PSSCRIPTANALYZER_SETTINGS_PATH) {
    $analyzeParams.Settings = $env:PSSCRIPTANALYZER_SETTINGS_PATH
}

# Run PSScriptAnalyzer
$issues   = Invoke-ScriptAnalyzer @analyzeParams
$errors   = ($issues.where({$_.Severity -eq 'Error'})).Count
$warnings = ($issues.where({$_.Severity -eq 'Warning'})).Count
$infos    = ($issues.where({$_.Severity -eq 'Information'})).Count

$strings = @{
    summary    = 'PSScriptAnalyzer results:{0}Errors: {1, 6}{2}Warnings: {3, 4}{4}Information: {5}'
    errorList  = '{0}The following PSScriptAnalyzer errors caused the check to fail:{1}'
    warningMsg = '{0} There were **[{1}]** warnings and **[{2}]** informational issues found. These did not cause the check to fail but it is recommended that they be fixed.'
}

# Create analysis summary
$summary = ($strings.summary -f $nl, $errors, $nl, $warnings, $nl, $infos)
$comment = '```' + $nl + $summary + $nl + '```'
if ($errors -gt 0) {
    $comment += $strings.errorList -f $nl, $nl
    $errorMsg = ($issues.Where({$_.Severity -eq 'Error'}) |
        Format-List -Property RuleName, Severity, ScriptName, Line, Message |
        Out-String -Width 80).Trim()
    $comment += '```' + $nl + $errorMsg + $nl + '```'
}
if (($warnings -gt 0) -or ($infos -gt 0)) {
    $comment += $strings.warningMsg -f $nl, $warnings, $infos
}
Write-Output $comment

# Send comment back to PR if any errors were found
$ghEvent = Get-Content -Path $env:GITHUB_EVENT_PATH | ConvertFrom-Json
if ($errors -gt 0) {
    $params = @{
        Uri = $ghEvent.pull_request.'_links'.comments.href
        Method = 'Post'
        Headers = @{
            Authorization = "token $env:GITHUB_TOKEN"
        }
        ContentType = 'application/json'
        Body = @{body = $comment} | ConvertTo-Json
    }
    Invoke-RestMethod @params > $null
}

exit $errors

entrypoint.ps1
[cmdletbinding()]
param()

$ErrorActionPreference = 'Stop'

$nl = [Environment]::NewLine

$analyzeParams = @{
    Recurse = $true
}

# By default, run PSScriptAnalyzer on the whole repository
# but allow overriding this with INPUT_ROOTPATH environment variable
if ($env:INPUT_ROOTPATH) {
    $analyzeParams.Path = Join-Path '/github/workspace' $env:INPUT_ROOTPATH
} else {
    $analyzeParams.Path = $env:GITHUB_WORKSPACE
}

# Path to custom script analyzer settings
if ($env:INPUT_SETTINGSPATH) {
    $analyzeParams.Settings = Join-Path '/github/workspace' $env:INPUT_SETTINGSPATH
}

# Run PSScriptAnalyzer
$issues   = Invoke-ScriptAnalyzer @analyzeParams
$errors   = $issues.Where({$_.Severity -eq 'Error'})
$warnings = $issues.Where({$_.Severity -eq 'Warning'})
$infos    = $issues.Where({$_.Severity -eq 'Information'})

# Create comment string
$comment  = '**PSScriptAnalyzer results:**'
$comment += '{0}<details><summary>Errors: [{1}], Warnings: [{2}], Information: [{3}]</summary><p>{4}{5}```' -f $nl, $errors.Count, $warnings.Count, $infos.Count, $nl, $nl
if ($errors.Count -gt 0) {
    $comment += $nl + ($errors | Format-List -Property RuleName, Severity, ScriptName, Line, Message | Out-String -Width 80).Trim()
}
if ($warnings.Count -gt 0) {
    $comment += $nl+ $nl + ($warnings | Format-List -Property RuleName, Severity, ScriptName, Line, Message | Out-String -Width 80).Trim()
}
if ($infos.Count -gt 0) {
    $comment += $nl + $nl + ($infos | Format-List -Property RuleName, Severity, ScriptName, Line, Message | Out-String -Width 80).Trim()
}
$comment += '{0}{1}```{2}</p></details>' -f $nl, $nl, $nl
Write-Output $comment

# Get comment URL
$ghEvent     = Get-Content -Path $env:GITHUB_EVENT_PATH | ConvertFrom-Json -Depth 30
$commentsUrl = $ghEvent.pull_request.comments_url

# Send comment back to PR if any issues were found
if ($commentsUrl -and ($env:INPUT_SENDCOMMENT -eq "$true" -or $env:INPUT_SENDCOMMENT -eq 1) -and $env:INPUT_REPOTOKEN -and ($errors.Count -gt 0 -or $warnings.Count -gt 0 -or $infos.Count -gt 0)) {
    $params = @{
        Uri = $commentsUrl
        Method = 'Post'
        Headers = @{
            Authorization = "token $env:INPUT_REPOTOKEN"
        }
        ContentType = 'application/json'
        Body = @{body = $comment} | ConvertTo-Json
    }
    Invoke-RestMethod @params > $null
}

$exitCode = 0
if ($env:INPUT_FAILONERRORS   -eq 'true' -or $env:INPUT_FAILONERRORS   -eq 1) { $exitCode += $errors.Count}
if ($env:INPUT_FAILONWARNINGS -eq 'true' -or $env:INPUT_FAILONWARNINGS -eq 1) { $exitCode += $warnings.Count}
if ($env:INPUT_FAILONINFOS    -eq 'true' -or $env:INPUT_FAILONINFOS    -eq 1) { $exitCode += $infos.Count}
exit $exitCode
